sdk/core/azure-core/azure/core/rest/_http_response_impl_async.py

```diff
@@ -0,0 +1,116 @@
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+from typing import AsyncIterator
+from ._rest_py3 import AsyncHttpResponse as _AsyncHttpResponse
+from ._http_response_impl import _HttpResponseBaseImpl
+
+class AsyncHttpResponseImpl(_HttpResponseBaseImpl, _AsyncHttpResponse):
+    """AsyncHttpResponseImpl built on top of our HttpResponse protocol class.
+
+    Since ~azure.core.rest.AsyncHttpResponse is an abstract base class, we need to
+    implement HttpResponse for each of our transports. This is an implementation
+    that each of the sync transport responses can be built on.
+
+    :keyword request: The request that led to the response
+    :type request: ~azure.core.rest.HttpRequest
+    :keyword any internal_response: The response we get directly from the transport. For example, for our requests
+     transport, this will be a requests.Response.
+    :keyword optional[int] block_size: The block size we are using in our transport
+    :keyword int status_code: The status code of the response
+    :keyword str reason: The HTTP reason
+    :keyword str content_type: The content type of the response
+    :keyword MutableMapping[str, str] headers: The response headers
+    :keyword Callable stream_download_generator: The stream download generator that we use to stream the response.
+    """
+
+    async def _set_read_checks(self):
+        self._is_stream_consumed = True
+        await self.close()
+
+    async def read(self) -> bytes:
+        """Read the response's bytes into memory.
+
+        :return: The response's bytes
+        :rtype: bytes
+        """
+        if self._content is None:
+            parts = []
+            async for part in self.iter_bytes():
+                parts.append(part)
+            self._content = b"".join(parts)
+        await self._set_read_checks()
+        return self._content
+
+    async def iter_raw(self) -> AsyncIterator[bytes]:
+        """Asynchronously iterates over the response's bytes. Will not decompress in the process
+        :return: An async iterator of bytes from the response
+        :rtype: AsyncIterator[bytes]
+        """
+        self._stream_download_check()
+        async for part in self._stream_download_generator(
+            response=self, pipeline=None, decompress=False
+        ):
+            yield part
+        await self.close()
+
+    async def iter_bytes(self) -> AsyncIterator[bytes]:
+        """Asynchronously iterates over the response's bytes. Will decompress in the process
+        :return: An async iterator of bytes from the response
+        :rtype: AsyncIterator[bytes]
+        """
+        if self._content is not None:
+            for i in range(0, len(self.content), self._block_size):
+                yield self.content[i : i + self._block_size]
+        else:
+            self._stream_download_check()
```

Comment:
is_stream_consumed indicates "Whether the stream has been fully consumed.", why you set it before the stream is all read?

Reply:
This is consistent [with httpx code](https://github.com/encode/httpx/blob/master/httpx/_models.py#L1627) but also, we can't quite set it after consuming the stream, because we're yielding the chunks from the response

Reply:
Yes. Then should we update the doc to "Whether the stream has been ~~fully~~ consumed."

Reply:
sounds fair

Reply:
updated!